src/mallocTracer.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "asprof.h"
+#include "assert.h"
+#include "codeCache.h"
+#include "mallocTracer.h"
+#include "os.h"
+#include "profiler.h"
+#include <dlfcn.h>
+#include <string.h>
+
+#define ADDRESS_OF(sym) ({               \
+    void* addr = dlsym(RTLD_NEXT, #sym); \
+    addr != NULL ? (sym##_t)addr : sym;  \
+})
+
+typedef void* (*malloc_t)(size_t);
+static malloc_t _orig_malloc = NULL;
+
+typedef void* (*calloc_t)(size_t, size_t);
+static calloc_t _orig_calloc = NULL;
+
+typedef void* (*realloc_t)(void*, size_t);
+static realloc_t _orig_realloc = NULL;
+
+typedef void (*free_t)(void*);
+static free_t _orig_free = NULL;
+
+static void* malloc_hook(size_t size) {
+    void* ret = _orig_calloc(1, size);
+    if (likely(ret && size)) {
+        MallocTracer::recordMalloc(ret, size, OS::nanotime());
+    }
+    return ret;
+}
+
+#ifdef __linux__
```

Comment:
Will macOS build break if you don't surround it with ifdefs?
Ultimately we want to support `nativemem` on macOS too.

Reply:
With that many linux-specific ifdefs, it's better to reorganize the code and put it in a platform-specific file.
But I would prefer just removing ifdefs if this does not do any harm on macOS.

Reply:
Removing ifdefs segfault on macos, adding back for now: https://github.com/async-profiler/async-profiler/actions/runs/12073081069/job/33668452329

There are 4 ifdefs in this file, only on the exported weak functions. I doubt the MacOS problem is in their implementation. We can keep them in the same file, ifdefs to be removed with mac support.